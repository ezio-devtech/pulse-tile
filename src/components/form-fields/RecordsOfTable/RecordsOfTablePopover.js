import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash/fp';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Spinner from '../../ui-elements/Spinner/Spinner';
import RecordsOfTablePopoverDiagnosis from './RecordsOfTablePopoverDiagnosis';
import RecordsOfTablePopoverMedications from './RecordsOfTablePopoverMedications';
import { fetchPatientDiagnosesDetailRequest } from '../../pages/Diagnosis/ducks/fetch-patient-diagnoses-detail.duck';
import { fetchPatientMedicationsDetailRequest } from '../../pages/Medications/ducks/fetch-patient-medications-detail.duck';
import { patientDiagnosesDetailSelector } from '../../pages/Diagnosis/selectors';
import { patientMedicationsDetailSelector } from '../../pages/Medications/selectors';

import RecordsOfTablePlugins from '../../theme/config/recordsOfTable/recordsOfTablePopover';
import { themeDetailsAction, themeDetailsSelector, themeTypesRecordsDetails } from '../../theme/config/recordsOfTable/recordsOfTableDetails';

const coreActions = {
    fetchPatientDiagnosesDetailRequest,
    fetchPatientMedicationsDetailRequest,
};
const actionsArray = Object.assign(coreActions, themeDetailsAction);
const mapDispatchToProps = dispatch => ({ actions: bindActionCreators(actionsArray, dispatch) });

const coreTypesRecordsDetails = {
  diagnosis: {
    title: 'Diagnosis',
    fetchDetail: 'fetchPatientDiagnosesDetailRequest',
    stateName: 'diagnosisDetail',
  },
  medications: {
    title: 'Medications',
    fetchDetail: 'fetchPatientMedicationsDetailRequest',
    stateName: 'medicationDetail',
  },
};
const typesRecordsDetails = Object.assign(coreTypesRecordsDetails, themeTypesRecordsDetails);

@connect(patientDiagnosesDetailSelector, mapDispatchToProps)
@connect(patientMedicationsDetailSelector)
@connect(themeDetailsSelector)

export default class RecordsOfTablePopover extends PureComponent {

  static propTypes = {
    record: PropTypes.object,
  };

  state = {
    typesRecords: typesRecordsDetails,
    sourceId: '',
  };

  componentDidMount() {
    const { actions, match, record: { type, sourceId } } = this.props;
    const { typesRecords } = this.state;
    const userId = _.get('params.userId', match);
    const fetchRequest = typesRecords[type] ? typesRecords[type].fetchDetail : '';
    if (fetchRequest && userId && sourceId) {
      this.setState({ sourceId });
      actions[fetchRequest]({ userId, sourceId })
    }
  }

  render() {
    const { record: { type: typeOfRecord } } = this.props;
    const { typesRecords, sourceId } = this.state;
    let title = '';
    let detail = null;
    if (typesRecords[typeOfRecord]) {
      title = typesRecords[typeOfRecord].title;
      detail = this.props[typesRecords[typeOfRecord].stateName] || null;
    }
    return (
      <div className="record-popover" style={{ display: 'block' }}>
        <div className="record-popover-header">
          <div className="record-popover-title">{title}</div>
        </div>
        <div className="record-popover-body">
          <div className="record-popover-content">
            {!detail || detail.sourceId !== sourceId ? <Spinner /> : null }
            {typeOfRecord === 'diagnosis' ? <RecordsOfTablePopoverDiagnosis detail={detail} /> : null}
            {typeOfRecord === 'medications' ? <RecordsOfTablePopoverMedications detail={detail} /> : null}
            <RecordsOfTablePlugins typeOfRecord={typeOfRecord} detail={detail} />
          </div>
        </div>
      </div>
    )
  }
}
