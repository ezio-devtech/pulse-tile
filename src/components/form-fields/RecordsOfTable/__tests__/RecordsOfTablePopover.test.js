import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import configureStore from 'redux-mock-store';

import RecordsOfTablePopover from '../../../form-fields/RecordsOfTable/RecordsOfTablePopover';

Enzyme.configure({ adapter: new Adapter() });

const userId = '9999999000';

const mockStore = configureStore();
const storeResource = {
  diagnosesDetail: {
    '9999999000': {
      problem: '# CONDYLE RIGHT HUMERUS 123',
      dateOfOnset: 1514851200000,
      description: 'ggg',
      terminology: 'SNOMED-CT',
      code: 12393890,
      author: 'Dr Tony Shannon',
      dateCreated: 1516801743000,
      source: 'ethercis',
      sourceId: 'a360e5fa-c837-4060-9f67-fecbf2f42d42',
      originalComposition: '',
      originalSource: '',
    },
  },
  medicationsDetail: {
    '9999999000': {
      name: 'test medication 11.01',
      doseAmount: 'Dose Amount',
      doseDirections: 'Dose Direction',
      doseTiming: 'Dose Amount',
      route: 'Po Per Oral',
      startDate: 1515719699000,
      startTime: 4499000,
      medicationCode: 123456789,
      medicationTerminology: 'external',
      author: 'bob.smith@gmail.com',
      dateCreated: 1515670672000,
      source: 'ethercis',
      sourceId: '1893365f-4961-4af1-847c-8ccf205aa7a1',
    },
  },
};
const store = mockStore(Object.assign({}, storeResource));
const emptyStore = mockStore({
  diagnosesDetail: {
    '9999999000': {},
  },
  medicationsDetail: {
    '9999999000': {},
  },
});


const DATE_TO_USE = new Date('2017');
const DATE_TO_USE_TIME = DATE_TO_USE.getTime();
global.Date = jest.fn(() => DATE_TO_USE);
Date.now = jest.fn(() => DATE_TO_USE_TIME);

const match = {
  params: {
    userId,
  },
};

describe('Component <RecordsOfTablePopover />', () => {

  it('should renders with props correctly when type diagnosis', () => {
    const component = shallow(
      <RecordsOfTablePopover
        store={store}
        record={
          {
            type: 'diagnosis',
            sourceId: '9dabaf87-7ad4-410a-944f-2449a8d0f8f7',
          }
        }
        match={match}
      />)
      .dive()
      .dive()
      .dive();

    expect(component.find('.record-popover')).toHaveLength(1);
    expect(component.find('.record-popover-title')).toHaveLength(1);
    expect(component.find('.record-popover-title').at(0).text()).toEqual('Diagnosis');
    expect(component.find('Spinner')).toHaveLength(1);
    expect(component.find('RecordsOfTablePopoverDiagnosis')).toHaveLength(1);

    expect(component).toMatchSnapshot();
  });

  it('should renders with props correctly when type medications', () => {
    const component = shallow(
      <RecordsOfTablePopover
        store={store}
        record={
          {
            type: 'medications',
            sourceId: '9dabaf87-7ad4-410a-944f-2449a8d0f8f7',
          }
        }
        match={match}
      />)
      .dive()
      .dive()
      .dive();

    expect(component.find('.record-popover')).toHaveLength(1);
    expect(component.find('.record-popover-title')).toHaveLength(1);
    expect(component.find('.record-popover-title').at(0).text()).toEqual('Medications');
    expect(component.find('Spinner')).toHaveLength(1);
    expect(component.find('RecordsOfTablePopoverMedications')).toHaveLength(1);

    expect(component).toMatchSnapshot();
  });

  it('should renders with props correctly without Spinner', () => {
    const component = shallow(
      <RecordsOfTablePopover
        store={emptyStore}
        record={
          {
            type: 'test',
            sourceId: '',
          }
        }
        match={match}
      />)
      .dive()
      .dive()
      .dive();

    expect(component).toMatchSnapshot();
  });

  it('should renders with props correctly without Spinner', () => {
    const component = shallow(
      <RecordsOfTablePopover
        store={emptyStore}
        record={
          {
            type: 'medications',
            sourceId: '',
          }
        }
        match={match}
      />)
      .dive()
      .dive()
      .dive();

    expect(component).toMatchSnapshot();
  });
});

