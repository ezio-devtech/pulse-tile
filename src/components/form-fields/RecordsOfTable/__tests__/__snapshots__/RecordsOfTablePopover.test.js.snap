// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <RecordsOfTablePopover /> should renders with props correctly when type diagnosis 1`] = `
<div
  className="record-popover"
  style={
    Object {
      "display": "block",
    }
  }
>
  <div
    className="record-popover-header"
  >
    <div
      className="record-popover-title"
    >
      Diagnosis
    </div>
  </div>
  <div
    className="record-popover-body"
  >
    <div
      className="record-popover-content"
    >
      <Spinner />
      <RecordsOfTablePopoverDiagnosis
        detail={
          Object {
            "author": "Dr Tony Shannon",
            "code": 12393890,
            "dateCreated": 1516801743000,
            "dateOfOnset": 1514851200000,
            "description": "ggg",
            "originalComposition": "",
            "originalSource": "",
            "problem": "# CONDYLE RIGHT HUMERUS 123",
            "source": "ethercis",
            "sourceId": "a360e5fa-c837-4060-9f67-fecbf2f42d42",
            "terminology": "SNOMED-CT",
          }
        }
      />
      <RecordsOfTablePlugins
        detail={
          Object {
            "author": "Dr Tony Shannon",
            "code": 12393890,
            "dateCreated": 1516801743000,
            "dateOfOnset": 1514851200000,
            "description": "ggg",
            "originalComposition": "",
            "originalSource": "",
            "problem": "# CONDYLE RIGHT HUMERUS 123",
            "source": "ethercis",
            "sourceId": "a360e5fa-c837-4060-9f67-fecbf2f42d42",
            "terminology": "SNOMED-CT",
          }
        }
        typeOfRecord="diagnosis"
      />
    </div>
  </div>
</div>
`;

exports[`Component <RecordsOfTablePopover /> should renders with props correctly when type medications 1`] = `
<div
  className="record-popover"
  style={
    Object {
      "display": "block",
    }
  }
>
  <div
    className="record-popover-header"
  >
    <div
      className="record-popover-title"
    >
      Medications
    </div>
  </div>
  <div
    className="record-popover-body"
  >
    <div
      className="record-popover-content"
    >
      <Spinner />
      <RecordsOfTablePopoverMedications
        detail={
          Object {
            "author": "bob.smith@gmail.com",
            "dateCreated": 1515670672000,
            "doseAmount": "Dose Amount",
            "doseDirections": "Dose Direction",
            "doseTiming": "Dose Amount",
            "medicationCode": 123456789,
            "medicationTerminology": "external",
            "name": "test medication 11.01",
            "route": "Po Per Oral",
            "source": "ethercis",
            "sourceId": "1893365f-4961-4af1-847c-8ccf205aa7a1",
            "startDate": 1515719699000,
            "startTime": 4499000,
          }
        }
      />
      <RecordsOfTablePlugins
        detail={
          Object {
            "author": "bob.smith@gmail.com",
            "dateCreated": 1515670672000,
            "doseAmount": "Dose Amount",
            "doseDirections": "Dose Direction",
            "doseTiming": "Dose Amount",
            "medicationCode": 123456789,
            "medicationTerminology": "external",
            "name": "test medication 11.01",
            "route": "Po Per Oral",
            "source": "ethercis",
            "sourceId": "1893365f-4961-4af1-847c-8ccf205aa7a1",
            "startDate": 1515719699000,
            "startTime": 4499000,
          }
        }
        typeOfRecord="medications"
      />
    </div>
  </div>
</div>
`;

exports[`Component <RecordsOfTablePopover /> should renders with props correctly without Spinner 1`] = `
<div
  className="record-popover"
  style={
    Object {
      "display": "block",
    }
  }
>
  <div
    className="record-popover-header"
  >
    <div
      className="record-popover-title"
    />
  </div>
  <div
    className="record-popover-body"
  >
    <div
      className="record-popover-content"
    >
      <Spinner />
      <RecordsOfTablePlugins
        detail={null}
        typeOfRecord="test"
      />
    </div>
  </div>
</div>
`;

exports[`Component <RecordsOfTablePopover /> should renders with props correctly without Spinner 2`] = `
<div
  className="record-popover"
  style={
    Object {
      "display": "block",
    }
  }
>
  <div
    className="record-popover-header"
  >
    <div
      className="record-popover-title"
    >
      Medications
    </div>
  </div>
  <div
    className="record-popover-body"
  >
    <div
      className="record-popover-content"
    >
      <Spinner />
      <RecordsOfTablePopoverMedications
        detail={null}
      />
      <RecordsOfTablePlugins
        detail={null}
        typeOfRecord="medications"
      />
    </div>
  </div>
</div>
`;
